k2/addons/k2-serial/zserial/serial.py

```diff
@@ -54,6 +55,75 @@ def serial_exec(sut, messagebus):
     return SerialClient(messagebus, sut.entity, sut.serial.timeout, sut.serial.prompt)
 
 
+@requires(sut='Sut', can=['serial'])
+@requires(messagebus='MessageBus')
+@component(name='RawSerialPort', can=['serial'], provided_by_extension='zserial')
+class RawSerialPort(object):
+    """
+    Access the serial port in raw byte mode.
+
+    Temporarily suspend the packetized log line handling to provide raw byte
+    access to the serial port.
+
```

Comment:
Add a warning that this mechanism is not thread-safe - in multiple aspects.
Suspension and resuming can be done by anyone. Which means that if multiple entities has used this feature overlapping in time, the first resume will destroy for the others.
Also, since multiple entities can suspend, the port is not guaranteed to be free. It can be used by other "suspenders".

Consider if the suspend-dispatcher should be blocking, and not return if it has been suspended by another context, until the service is resumed. Alternatively, that it raises an exception if you try to suspend if already suspended.
The minimum would at least to warn about it in the documentation. If you can find a nice solution for it is a plus.

Reply:
Agreed. I'll see what I can do about the thread safety

Reply:
I added exceptions for now. I don't know if blocking is the way to go. E.g. transferring a large file @ 9600 baud would block someone else for a really long time. Maybe that's ok, but maybe that will trigger a timeout exception in some other (future) component waiting to suspend, in which case it would have been easier to troubleshoot if the SerialExtension had thrown an exception right away? :shrug:
These are all hypothetical scenarios. I suggest we leave it at throwing exceptions when trying to suspend when already suspended and when trying to resume when not suspended.

Reply:
I consider throwing an exception good enough, so I'm resolving this conversation :slightly_smiling_face: